execute/tokendata/observer.go

```diff
@@ -2,42 +2,145 @@ package tokendata
 
 import (
 	"context"
+	"errors"
 
 	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
 
 	"github.com/smartcontractkit/chainlink-ccip/execute/exectypes"
+	"github.com/smartcontractkit/chainlink-ccip/execute/tokendata/usdc"
+	"github.com/smartcontractkit/chainlink-ccip/pluginconfig"
 )
 
 type TokenDataObserver interface {
+	// Observe takes the message observations and returns token data observations for matching tokens.
+	// If TokenDataObserver doesn't support the token it must return exectypes.NotSupportedTokenData.
+	// This way higher layer will know how to orchestrate Observers and which observations are relevant.
+	//
+	// TokenDataObserver must return data for every token that is present in the exectypes.MessageObservations.
+	// Meaning that exectypes.TokenDataObservations must reflect the structure of exectypes.MessageObservations
+	// (the same number of objects and tokens).
+	// * if token is supported and requires offchain processing, TokenDataObserver must initialize the token according
+	// 	 to its logic. Depending on the result it must initialize with either exectypes.NewSuccessTokenData or
+	//	 exectypes.NewErrorTokenData
+	// * if token is not supported, TokenDataObserver must set the token data to exectypes.NotSupportedTokenData
 	Observe(
 		ctx context.Context,
 		observations exectypes.MessageObservations,
 	) (exectypes.TokenDataObservations, error)
+
+	// IsTokenSupported returns true if the token is supported by the observer.
+	IsTokenSupported(sourceChain cciptypes.ChainSelector, msgToken cciptypes.RampTokenAmount) bool
 }
 
 // CompositeTokenDataObserver is a TokenDataObserver that combines multiple TokenDataObserver behind the same interface.
 // Goal of that is to support multiple token observers supporting different tokens (e.g. CCTP, MyFancyToken etc)
 type CompositeTokenDataObserver struct {
-	Observers []TokenDataObserver
+	observers []TokenDataObserver
+}
+
+// nolint early-return
+func NewConfigBasedCompositeObservers(config []pluginconfig.TokenDataObserverConfig) (*CompositeTokenDataObserver, error) {
+	observers := make([]TokenDataObserver, len(config))
+	for i, c := range config {
```

Comment:
Is there an assumption here that `config` is de-duped, i.e we don't have something like `[usdcConfig, usdcConfig]`?

Reply:
At this stage we assume config is deduped, I will add the test to the config parser to cover that edge case